src/tike/operators/cupy/convolution.cu

```diff
@@ -7,9 +7,9 @@
 // Padding areas are untouched and retain whatever values they had before the
```

Comment:
@carterbox 1) why kernels have so many loops inside for such a simple operation of extracting patches? It is not clear what they are computing. Why did you decide to stop using this code: https://github.com/nikitinvv/libtike-cufft/blob/master/src/cuda/kernels.cu ? 2) It is better to have the block size as 1024 (or 512,256 if the kernel is heavy and uses many registers.)

Reply:
In my commit message, I explained that this kernel has been converted into a grid-stride-loop structure. This replaces the `if (thread index too high) return` statements with loop conditions.

Grid-stride-loops allow flexible kernel launches because you can run them with any number of threads. For example, you can run them with a single thread when debugging. However, their main purpose is you can run them when there are not enough threads available to index the entire dataset (that's why they are called grid-stride). This also means you can *intentionally* reduce the number of kernel launches.

I haven't actually tuned the grid and block sizes to do this, but in theory, having fewer long running threads reduces the overhead from thread creation and distruction.

I read about these loops on the NVidia developer blog, you can read that post for more information.

Reply:
@carterbox I am fine with this code because it does not influence the performance much.

But the explanation for the main purpose of grid-stride loops is not true. 
Max dimension size of a thread block (x,y,z): (1024, 1024, 64)
Max dimension size of a grid size    (x,y,z): (2147483647, 65535, 65535)
So max global thread id is 1023*2147483647+1023*65535+63*65535+63 which is definitely enough to process the entire dataset. With a standard settings for cuda kernels:
  const int tx = blockDim.x * blockIdx.x + threadIdx.x;
  const int ty = blockDim.y * blockIdx.y + threadIdx.y;
  const int tz = blockDim.z * blockIdx.z + threadIdx.z;
you can index any dataset that we aim to reconstruct, and the code looks simpler.

"
I haven't actually tuned the grid and block sizes to do this, but in theory, having fewer long running threads reduces the overhead from thread creation and destruction.
"
At the same time, this may lead to not optimal conveyor processing. The cuda scheduler switches threads blocks in an optimal way. If some blocks are on the stage of reading or writing to memory, then the scheduler can switch to other blocks that need computations. With a small number of threads there might be not enough blocks to switch.

The complexity of kernel launches depends on the number of registers and shared memory used. This kernel is very light in this sense, so more threads should be better.

Reply:
@nikitinvv, you're estimating the number of threads incorrectly? Although the maximum sizes of the dimensions of a block are (1024, 1024, 64), the total number of threads in a block still must remain less than or equal to 1024 [1]. i.e. (1024, 1, 1) is allowed (1024, 2, 1) is not allowed.

Therefore, the maximum global thread count is 1024 * (2147483647 * 65535 * 65535). Or with your previous thread mapping:
```
tx = 1024 * 2147483647  # probe
ty = 65535  # scan
tz = 65535  # theta
```
I guess 6.5e4 scan positions is enough? I'm not sure a dataset with more scan positions would fit into the memory of 1 GPU. Unless we are using unified memory and trying to process the 2.2e7 scan positions on 1 GPU.




[1] [Table 15.](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#features-and-technical-specifications)

Reply:
I'm open to playing with this kernel more in the future, but for now, I'm going to merge to unblock xiaodong.